### PR TITLE
2693 cat2 messaging cleanup

### LINKED ISSUE #2693: [Error Audit] Category 2 error messages clean-up
**Description**
Delivers Category 2 error language fixes 

**Acceptance Criteria**:
- [x] All Cat2 errors should include the `[Item #][Friendly Name]` pattern (see `run_field_validators` in row schema)
- [ ] All Cat 2 errors are consistent with desired outputs in the [HackMD table](https://tdp-project.gitbook.io/project-trackers-and-guides#cat-2-validators)
- [ ]  Misc cleanups are fixed:
   - [ ] `validateRace` error message typo fix Desired output: `f"{value} is not greater than or equal to 0 or smaller than or equal to 2.`

**Tasks**:
- [ ] Correct row schema to add Item/Friendly names to Cat2 errors
- [ ] Correct misc cleanup re: `validateRace`
- [ ] Correct other issues noted in Hack.md table which includes validators `oneOf, notOneOf, isInStringRange, isNumber`

**Supporting Documentation**:

- [Error Audit HackMD](https://tdp-project.gitbook.io/project-trackers-and-guides#cat-2-validators)

### PR DESCRIPTION
**Pull request by jtimpe**
## Summary of Changes
Pull request closes #2693
* implements the `{record_type} Item {item_num} {friendly_name}: {error_msg}` format for all cat2 validators
    * generic `format_error_context` function makes future changes consistent
    * note: this makes cat3 if/then/or validators look a little funny -> will create spike for research
* Omit all single quotes to enhance readability of values when displaying list of `options`
* Strip leading/trailing spaces from values in `isNumber`
* Not addressed: We might have a little bit of a broader issue here with fields that are not required serving up errors saying they're required. If it's a space fill issue it should not produce an error indicating that it's a missing value. (referring to a cat3 validator, would be addressed by #2821)
* Not addressed: `validateRace` error message typo fix - I didn't see a difference between the message included in the AC and what's already in the repo. If i missed something, please let me know

## How to Test
```
cd tdrs-backend && docker-compose up
cd tdrs-frontend && docker-compose up --build
```

1. Open http://localhost:3000/ and sign in.
1. Upload data files containing errors
1. Inspect the generated error report 

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] Correct row schema to add Item/Friendly names to Cat2 errors
+ [x] Correct other issues noted in Hack.md table which includes validators oneOf, notOneOf, isInStringRange, isNumber
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/parsers/validators.py [hunk 1]
```diff
@@ -217,78 +217,111 @@ def sumIsLargerFunc(value, row_schema):
     return sumIsLargerFunc
 
 
+def recordHasLength(length):
```
Comment (jtimpe) [inline tdrs-backend/tdpservice/parsers/validators.py:221]: i moved the following functions for clarity, but haven't modified them. they are used as preparsing validators, so we pass `0` in for the field number, and `-1` for item number. this results in each error being prefixed with `{record_type} Item -1 record type: {error message}`, where `record type` is not the field related to the error. I left the messages un-touched for clarity but open to suggestions on how to address this

**Review by elipe17 — APPROVED** [review]
I have one question, but it doesn't halt the PR. In the Errors audit, @reitermb has the friendly_name in parenthesis; e.g. `Item 26 (Number of Months Countable toward Federal Time Limit)`. I leave it up to him whether that should be updated in this PR. Otherwise, great work!

File: tdrs-backend/tdpservice/parsers/validators.py (hunk 1 above)
Comment (raftmsohani) [reply tdrs-backend/tdpservice/parsers/validators.py:221]: Maybe assign ``` PREPARSER = -1 ``` then check if preparser error in the validator =>``` record_type="preparser error"```

**Review by raftmsohani — APPROVED** [review]

**Comment by reitermb** [comment]
> I have one question, but it doesn't halt the PR. In the Errors audit, @reitermb has the friendly_name in parenthesis; e.g. `Item 26 (Number of Months Countable toward Federal Time Limit)`. I leave it up to him whether that should be updated in this PR. Otherwise, great work!

Yeah, all friendly names should be in parentheses next to the item #'s—I believe that's currently the case for lower categories, right?

I think with that this is good to go re: my review!

**Review by reitermb — APPROVED** [review]
w/ parentheses around friendly names this should be good to go

**Comment by ADPennington** [comment]
per sync with @jtimpe and @reitermb:

- I will continue reviewing this PR. This will not be an exhaustive test of all possible cat 2 error messages. 
- If qasp approved, need to scope a more thorough test of this PR in the `develop` branch environment. also need to discuss plan with @andrew-jameson for how/when to release this PR given cat 3 error message implications
- @jtimpe creating a spike ticket to explore options for potentially resolving at least some of the cat 3 error message clean-up needed as a result of this PR
- @reitermb and I still plan to discuss cat 3 error audit at UX syncs.

**Comment by ADPennington** [comment]
@jtimpe sorry just noticed merge conflicts on this

File: tdrs-backend/tdpservice/parsers/validators.py [hunk 2]
```diff
@@ -332,49 +366,52 @@ def startsWith(substring, error_func=None):
...
+        else f"{format_error_context(row_schema, friendly_name, item_num)}: {value} does not start with {substring}.",
     )
 
 
 def isNumber():
     """Validate that value can be casted to a number."""
     return make_validator(
-        lambda value: value.isnumeric(),
-        lambda value, row_schema, friendly_name, item_num: f"{row_schema.record_type}: {value} is not a number."
+        lambda value: str(value).strip().isnumeric(),
```
Comment (ADPennington) [inline tdrs-backend/tdpservice/parsers/validators.py:388]: noting for team awareness...it looks like this change was made to suppress error messages associated with leading spaces on fields like county fips code, and zip code. i thought there was another fix to address this in #2951. @jtimpe @elipe17 @reitermb

**Comment by ADPennington** [comment]
@jtimpe @reitermb @andrew-jameson i tested this PR based on the most common cat 1, cat2, and cat 3 errors in prod since sprint 98 was released to prod.  Please note:

- the error messages for cat 1 are in good shape
- some of the cat 2 error messages may need refinement:
   - the most common error message here is actually the one implemented in #2818. not sure if we want to re-consider making this a pre-parsing error or hard-code the error message. 
   -  there does not appear to be any changes to the error messages for legitimately required fields  that are space-filled like DOB and SSN. 
   - the change to the isnumeric validator may be fine but called it out because im not sure how it interacts with #2951 yet. 
- cat 3 error messages are confirmed to be impacted by this PR, as expected. 
- cat 4 error messages not impacted since these are currently hard-coded. 

Further below are the error messages I tested for reference ⬇️. let me know how we should proceed, especially re: cat 2 error messages. 

----
----
  - cat 1: pre-parsing errors
    
    - header 
      - incorrect length :heavy_check_mark: 
        - before: `HEADER record length is 24 characters but must be 23.` 
        - this PR: `HEADER: record length is 24 characters but must be 23.`
      - update indicator (technically cat 2 error in codebase) :warning: 
        - before: `HEADER update indicator: N does not match D.` 
        - this PR: `HEADER Item 10 (update indicator): N does not match D.` 
          - **feedback**: update indicator is not associated with an item number.  might be worth hard-coding the message or making this a pre-parsing validator since it blocks parsing.
      - missing header :heavy_check_mark: 
          - before:`HEADER record length is 2 characters but must be 23.
                Your file does not begin with a HEADER record.`
          - this PR: `HEADER: record length is 2 characters but must be 23.
                  Your file does not begin with a HEADER record.`
      
    - inconsistent fiscal period :heavy_check_mark: 
      - before:`Submitted reporting year:2022, quarter:Q4 doesn't match file reporting year:2023, quarter:Q4.`
      - this PR: `Submitted reporting year:2022, quarter:Q4 doesn't match file reporting year:2023, quarter:Q4.`
    - inconsistent program type :heavy_check_mark: 
      - Before:`Data does not match the expected layout for SSP Active Case Data.` 
      - this PR: `Data does not match the expected layout for SSP Active Case Data.` 
    - record not correct length :heavy_check_mark: 
      - before: `T1 record length is 116 characters but must be 156. 
                 No records created.`
      - this PR: `T1: record length of 116 characters is not in the range [117, 156].
                  No records created.`

   - cat 2 errors
     - before:
         - `T2: 14 is not in ['01', '02', '05', '07', '09', '15', '16', '17', '18', '99'].`
         - `date of birth is required but a value was not provided. `
         - `social security number is required but a value was not provided.`
         - T1: 45 is not a number.
     - after:
         - `T2 Item 49 (work participation status): 14 is not in [01, 02, 05, 07, 09, 15, 16, 17, 18, 99].` 
         -  `date of birth is required but a value was not provided. ` :warning: 
            -  **feedback**: required field doesnt reference item number. 
         - `social security number is required but a value was not provided.` :warning:
           - **feedback**: required field doesnt reference item number. 
         - no error message raised for county fips value = ` 45` :question: 
           - **feedback**: is this something to address here or in #2951
   - cat 3 errors
     - before:
       - `if family affiliation :1 validator1 passed then work participation status T2: 14 is not in ['01', '02', '05', '07', '09', '15', '17', '18', '19', '99'].` 
       - `if family affiliation :1 validator1 passed then social security number T2: 000000000 is in ['000000000', '111111111', '222222222', '333333333', '444444444', '555555555', '666666666', '777777777', '888888888', '999999999'].`
     - after: 
       - `if family affiliation :1 validator1 passed then work participation status T2 Item -1 (work participation status): 14 is not in [01, 02, 05, 07, 09, 15, 17, 18, 19, 99].` :warning: 
       - `if family affiliation :1 validator1 passed then social security number T2 Item -1 (social security number): 000000000 is in ['000000000', '111111111', '222222222', '333333333', '444444444', '555555555', '666666666', '777777777', '888888888', '999999999'].` :warning:
         -  **feedback**: incorrect item number reference due to conditional validator

File: tdrs-backend/tdpservice/parsers/validators.py (hunk 2 above)
Comment (jtimpe) [reply tdrs-backend/tdpservice/parsers/validators.py:388]: one thing to add - `isNumber` is not consistently used for the `COUNTY_FIPS_CODE` field validators across the different schemas - some use `isInStringRange` and `matches` as well. so this change may not be present in every instance we need it

**Comment by jtimpe** [comment]
@ADPennington i've updated the `is required` validator message to use the schema/item/friendly string formatting. this change applies to every field where the `required=True`

**Review by ADPennington — APPROVED** [review]
@jtimpe - lgtm; please note:

- the required field error message resolved for fields other than SSN and DOB (#3059). example: `T1 Item 8 (funding stream): field is required but a value was not provided.` 
- county fips issue was handled by #2951 
- cat 3-related issues persist and i will look into how its addressed in #3016 
- header cat 2 error will be addressed in #3049

**Pull request merged**